File: libs/cpp/capture_core/src/framing.cpp
// SPDX-License-Identifier: GPL-3.0-only
#include "capture/framing.hpp"

#include <cstring>
#include <stdexcept>

namespace capture {
namespace {

void write_u32_le(uint8_t* dst, uint32_t value) {
  dst[0] = static_cast<uint8_t>(value & 0xFF);
  dst[1] = static_cast<uint8_t>((value >> 8) & 0xFF);
  dst[2] = static_cast<uint8_t>((value >> 16) & 0xFF);
  dst[3] = static_cast<uint8_t>((value >> 24) & 0xFF);
}

uint32_t read_u32_le(const uint8_t* src) {
  return static_cast<uint32_t>(src[0]) |
         (static_cast<uint32_t>(src[1]) << 8) |
         (static_cast<uint32_t>(src[2]) << 16) |
         (static_cast<uint32_t>(src[3]) << 24);
}

}  // namespace

std::vector<uint8_t> encode_frame(uint32_t message_type,
                                  std::span<const uint8_t> payload,
                                  uint32_t correlation_id) {
  if (payload.size() > kMaxPayloadLen) {
    throw FrameError("payload exceeds max length");
  }
  std::vector<uint8_t> out(kFrameHeaderSize + payload.size());
  write_u32_le(out.data() + 0, kFrameMagic);
  write_u32_le(out.data() + 4, static_cast<uint32_t>(payload.size()));
  write_u32_le(out.data() + 8, message_type);
  write_u32_le(out.data() + 12, correlation_id);
  if (!payload.empty()) {
    std::memcpy(out.data() + kFrameHeaderSize, payload.data(), payload.size());
  }
  return out;
}

std::size_t decode_frame(std::span<const uint8_t> buffer, Frame& out) {
  if (buffer.size() < kFrameHeaderSize) {
    throw std::runtime_error("incomplete header");
  }
  const uint32_t magic = read_u32_le(buffer.data());
  if (magic != kFrameMagic) {
    throw FrameError("bad magic");
  }
  const uint32_t payload_len = read_u32_le(buffer.data() + 4);
  if (payload_len > kMaxPayloadLen) {
    throw FrameError("payload_len exceeds max");
  }
  const std::size_t total = kFrameHeaderSize + payload_len;
  if (buffer.size() < total) {
    throw std::runtime_error("incomplete payload");
  }
  out.message_type = read_u32_le(buffer.data() + 8);
  out.correlation_id = read_u32_le(buffer.data() + 12);
  out.payload.assign(buffer.begin() + static_cast<std::ptrdiff_t>(kFrameHeaderSize),
                    buffer.begin() + static_cast<std::ptrdiff_t>(total));
  return total;
}

void FrameDecoder::feed(std::span<const uint8_t> data) {
  buf_.insert(buf_.end(), data.begin(), data.end());
}

bool FrameDecoder::pop(Frame& out) {
  try {
    const std::size_t consumed = decode_frame(buf_, out);
    buf_.erase(buf_.begin(),
               buf_.begin() + static_cast<std::ptrdiff_t>(consumed));
    return true;
  } catch (const std::runtime_error& ex) {
    const std::string msg = ex.what();
    if (msg.find("incomplete") != std::string::npos) {
      return false;
    }
    throw;
  }
}

void FrameDecoder::reset() { buf_.clear(); }

}  // namespace capture

File: libs/python/capture_protocol/capture_protocol/framing.py
# SPDX-License-Identifier: Apache-2.0
"""CSP1 length-prefixed protobuf framing."""

from __future__ import annotations

import struct
from dataclasses import dataclass

from capture_protocol.constants import (
    FRAME_HEADER_SIZE,
    FRAME_MAGIC,
    FRAME_MAGIC_BYTES,
    MAX_PAYLOAD_LEN,
)

_HEADER_STRUCT = struct.Struct("<IIII")  # magic, payload_len, message_type, correlation_id


class FrameError(ValueError):
    """Invalid frame on the wire."""


@dataclass(frozen=True, slots=True)
class Frame:
    message_type: int
    correlation_id: int
    payload: bytes

    @property
    def is_event(self) -> bool:
        return self.correlation_id == 0


def encode_frame(message_type: int, payload: bytes, correlation_id: int = 0) -> bytes:
    if len(payload) > MAX_PAYLOAD_LEN:
        raise FrameError(f"payload_len {len(payload)} exceeds max {MAX_PAYLOAD_LEN}")
    header = _HEADER_STRUCT.pack(FRAME_MAGIC, len(payload), message_type, correlation_id)
    return header + payload


def decode_frame(buffer: bytes | bytearray | memoryview) -> tuple[Frame, int]:
    """Decode one frame from the start of buffer.

    Returns (frame, bytes_consumed).
    Raises FrameError on bad magic or oversized length.
    Raises ValueError if buffer is incomplete (need more data).
    """
    view = memoryview(buffer)
    if len(view) < FRAME_HEADER_SIZE:
        raise ValueError("incomplete header")
    magic, payload_len, message_type, correlation_id = _HEADER_STRUCT.unpack_from(view, 0)
    if magic != FRAME_MAGIC:
        # Helpful diagnostic: show what we got
        got = bytes(view[:4])
        raise FrameError(f"bad magic: expected {FRAME_MAGIC_BYTES!r}, got {got!r}")
    if payload_len > MAX_PAYLOAD_LEN:
        raise FrameError(f"payload_len {payload_len} exceeds max {MAX_PAYLOAD_LEN}")
    total = FRAME_HEADER_SIZE + payload_len
    if len(view) < total:
        raise ValueError("incomplete payload")
    payload = bytes(view[FRAME_HEADER_SIZE:total])
    return Frame(message_type, correlation_id, payload), total


class FrameDecoder:
    """Incremental byte-stream decoder for named-pipe / socket reads."""

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, data: bytes) -> list[Frame]:
        self._buf.extend(data)
        frames: list[Frame] = []
        while True:
            try:
                frame, consumed = decode_frame(self._buf)
            except ValueError:
                break
            frames.append(frame)
            del self._buf[:consumed]
        return frames

    def reset(self) -> None:
        self._buf.clear()
